src/ChunkGeneratorThread.h

```diff
@@ -0,0 +1,162 @@
+	class cChunkSink
+	{
+	public:
+		// Force a virtual destructor
+		virtual ~cChunkSink() {}
+
+		/** Called after the chunk has been generated
+		The interface may store the chunk, send it over network, whatever.
+		The chunk is not expected to be modified, but the generator will survive if the implementation
+		changes the data within. All changes are ignored, though. */
+		virtual void OnChunkGenerated(cChunkDesc & a_ChunkDesc) = 0;
+
+		/** Called just before the chunk generation is started,
+		to verify that it hasn't been generated in the meantime.
+		If this callback returns true, the chunk is not generated. */
+		virtual bool IsChunkValid(cChunkCoords a_Coords) = 0;
+
+		/** Called when the generator is overloaded to skip chunks that are no longer needed.
+		If this callback returns false, the chunk is not generated. */
+		virtual bool HasChunkAnyClients(cChunkCoords a_Coords) = 0;
+
+		/** Called to check whether the specified chunk is in the queued state.
+		Currently used only in Debug-mode asserts. */
+		virtual bool IsChunkQueued(cChunkCoords a_Coords) = 0;
+	} ;
+
+
+	cChunkGeneratorThread (void);
+	virtual ~cChunkGeneratorThread() override;
+
+	/** Read settings from the ini file and initialize in preperation for being started. */
+	bool Initialize(cPluginInterface & a_PluginInterface, cChunkSink & a_ChunkSink, cIniFile & a_IniFile);
+
+	void Stop(void);
+
+	/** Queues the chunk for generation
+	If a-ForceGenerate is set, the chunk is regenerated even if the data is already present in the chunksink.
+	a_Callback is called after the chunk is generated. If the chunk was already present, the callback is still called, even if not regenerating.
+	It is legal to set the callback to nullptr, no callback is called then.
+	If the generator becomes overloaded and skips this chunk, the callback is still called. */
+	void QueueGenerateChunk(cChunkCoords a_Coords, bool a_ForceRegeneration, cChunkCoordCallback * a_Callback = nullptr);
+
+	/** Generates the biomes for the specified chunk (directly, not in a separate thread). Used by the world loader if biomes failed loading. */
+	void GenerateBiomes(cChunkCoords a_Coords, cChunkDef::BiomeMap & a_BiomeMap);
+
+	void WaitForQueueEmpty();
+
+	int GetQueueLength();
+
+	int GetSeed() const { return m_Seed; }
+
+	/** Returns the biome at the specified coords. Used by ChunkMap if an invalid chunk is queried for biome */
+	EMCSBiome GetBiomeAt(int a_BlockX, int a_BlockZ);
+
+
+private:
+
+	struct QueueItem
+	{
+		/** The chunk coords */
+		cChunkCoords m_Coords;
+
+		/** Force the regeneration of an already existing chunk */
+		bool m_ForceRegeneration;
+
+		/** Callback to call after generating. */
+		cChunkCoordCallback * m_Callback;
+
+		QueueItem(cChunkCoords a_Coords, bool a_ForceRegeneration, cChunkCoordCallback * a_Callback):
+			m_Coords(a_Coords),
+			m_ForceRegeneration(a_ForceRegeneration),
+			m_Callback(a_Callback)
+		{
+		}
+	};
+
+	using Queue = std::list<QueueItem>;
```

Comment:
Wouldn't it make more sense to use a `std::queue` instead?

Reply:
Or the queue from `OSSupport/Queue.h` which should be thread-safe already.

Reply:
Again, code that has been just moved around.
`std::queue` is just a wrapper over `std::deque` by default.

OSSupport/Queue.h might be interesting, didn't remember we had one. I'll have a look at it.

Reply:
I has a look at the `cQueue`, but in the end, `cChunkGeneratorThread` does a bit more than just read and write to a queue, it calculates the throughput, and I'm too lazy to adapt that to use the `cQueue`, it isn't a straightforward change.